runtime/src/main/cpp/CyclicCollector.cpp

```diff
@@ -0,0 +1,435 @@
+    CHECK_CALL(pthread_cond_init(&cond_, nullptr), "Cannot init collector condition")
+    CHECK_CALL(pthread_create(&gcThread_, nullptr, gcWorkerRoutine, this), "Cannot start collector thread")
+  }
+
+  ~CyclicCollector() {
+    {
+      Locker locker(&lock_);
+      terminateCollector_ = true;
+      shallRunCollector_ = true;
+      CHECK_CALL(pthread_cond_signal(&cond_), "Cannot signal collector")
+    }
+    // TODO: improve waiting for collector termination.
+    while (atomicGet(&terminateCollector_)) {}
+    releasePendingUnlocked(nullptr);
+    pthread_cond_destroy(&cond_);
+    pthread_mutex_destroy(&lock_);
+  }
+
+  static void* gcWorkerRoutine(void* argument) {
+    CyclicCollector* thiz = reinterpret_cast<CyclicCollector*>(argument);
+    thiz->gcProcessor();
+    return nullptr;
+  }
+
+  void gcProcessor() {
+     {
+       Locker locker(&lock_);
+       KStdDeque<ObjHeader*> toVisit;
+       KStdUnorderedSet<ObjHeader*> visited;
+       int restartCount = 0;
+       while (!terminateCollector_) {
+         CHECK_CALL(pthread_cond_wait(&cond_, &lock_), "Cannot wait collector condition")
+         if (!shallRunCollector_) continue;
+         atomicSet(&gcRunning_, 1);
+         restartCount = 0;
+        restart:
+         if (restartCount > 10) {
+           COLLECTOR_LOG("wait for some time to avoid GC trashing\n");
+           struct timeval tv;
+           struct timespec ts;
+           long long nsDelta = 1000LL * 1000LL;
+           ts.tv_nsec = (tv.tv_usec * 1000LL + nsDelta) % 1000000000LL;
+           ts.tv_sec = (tv.tv_sec * 1000000000LL + nsDelta) / 1000000000LL;
+           pthread_cond_timedwait(&cond_, &lock_, &ts);
+         }
+         atomicSet(&mutatedAtomics_, 0);
+         visited.clear();
+         toVisit.clear();
+         sideRefCounts_.clear();
+         for (auto* root: rootset_) {
+           // We only care about frozen values here, as only they could become part of shared cycles.
+           if (!root->container()->frozen()) continue;
+           COLLECTOR_LOG("process root %p\n", root);
+           traverseObjectFields(root, [&toVisit, &visited](ObjHeader** location) {
+             ObjHeader* ref = *location;
+             if (ref != nullptr && visited.count(ref) == 0) {
+               toVisit.push_back(ref);
+               COLLECTOR_LOG("adding %p for visiting\n", ref);
+             }
+           });
+           if (atomicGet(&mutatedAtomics_) != 0) {
+             COLLECTOR_LOG("restarted during visiting collect\n")
+             restartCount++;
+             goto restart;
+           }
+           while (toVisit.size() > 0)  {
+             auto* obj = toVisit.front();
+             toVisit.pop_front();
+             COLLECTOR_LOG("visit %p\n", obj);
+             auto* objContainer = obj->container();
+             if (objContainer == nullptr) continue;  // Permanent object.
+             if (!objContainer->frozen()) continue;
```

Comment:
Can `objContainer->frozen()` be actually `false`? If not, then this should be an assertion.